### PR TITLE
Add new models for malware detection.

### LINKED ISSUE #7090: Create new models and tables for automated malware detection
This issue is part of the work defined by [2019-Q4-PyPI Milestone 2](https://github.com/python/request-for/blob/master/2019-Q4-PyPI/RFP.md#milestone-2---systems-for-automated-detection-of-malicious-uploads).

Create the check and verdict models, and write the corresponding data migrations to create these tables in the database.

### PR DESCRIPTION
**Pull request by xmunoz**
Fixes #7090 and #7092.

### CODE REVIEW

File: warehouse/malware/models.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+class MalwareVerdict(db.Model):
+    __tablename__ = "malware_verdicts"
+
+    run_date = Column(DateTime, nullable=False, server_default=sql.func.now())
+    check_id = Column(
+        ForeignKey("malware_checks.id", onupdate="CASCADE", ondelete="CASCADE"),
+        nullable=False,
+        index=True,
+    )
+    file_md5 = Column(
```
[the remark this answers is not in the document]
Comment (xmunoz) [reply warehouse/malware/models.py]: Two files could have the same content (same md5) but different ids. These would still be the same file, and the same verdicts would apply to them. There are some potential efficiency gains if we consider only the md5 of files. For example, I think it makes sense to only run a check against a file once. If we consider file IDs, then we will end of with a lot of redundant check verdicts.

File: warehouse/malware/models.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    name = Column(CIText, nullable=False)
+    version = Column(Integer, default=0, nullable=False)
+    short_description = Column(String(length=128), nullable=False)
+    long_description = Column(Text, nullable=False)
+    check_type = Column(
+        Enum(MalwareCheckTypes, values_callable=lambda x: [e.value for e in x]),
+        nullable=False,
+    )
+    # This field contains the same content as the ProjectEvent and UserEvent "tag"
```
[the remark this answers is not in the document]
Comment (xmunoz) [reply warehouse/malware/models.py]: That is correct. My understanding is also that they are free-form. I'm happy to add an issue for creating  a consistent list of these tags. Do you have any recommendations for how I could clarify this comment in the mean time?

File: warehouse/malware/models.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    )
+    classification = Column(
+        Enum(VerdictClassification, values_callable=lambda x: [e.value for e in x]),
+        nullable=False,
+    )
+    confidence = Column(
+        Enum(VerdictConfidence, values_callable=lambda x: [e.value for e in x]),
+        nullable=False,
+    )
+    message = Column(String(length=100), nullable=True)
```
[the remark this answers is not in the document]
Comment (xmunoz) [reply warehouse/malware/models.py]: Sounds good.

File: warehouse/malware/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ewdurbin) [reply warehouse/malware/models.py]: Ah, I understand what you're going for, but in the case of warehouse/PyPI all of the digests are [enforced to be unique](https://github.com/pypa/warehouse/blob/de6d014f62021cb915187f75458fa3ce53d115a7/warehouse/packaging/models.py#L503-L505), basically PyPI doesn't allow uploading the same file with a different name or to a different project.

Given that it sounds like relating via the object ID is probably OK?

[the remark this answers is not in the document]
Comment (dstufft) [reply warehouse/malware/models.py]: I would not rely on MD5 for deduplication of file checks, I'd use one of the stronger hashes but I think that you can get deduplication regardless of what the FK is. If I recall correctly string based FKs perform a lot more poorly than UUID/Int based FKs, but I could be misremembering that.

[the remark this answers is not in the document]
Comment (xmunoz) [reply warehouse/malware/models.py]: Ah, good to know @ewdurbin and @dstufft! Yeah, I overlooked the uniqueness constraint, and performance is definitely worse on strings. I'll make the change to ID.

File: warehouse/malware/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ewdurbin) [reply warehouse/malware/models.py]: [I filed an issue](https://github.com/pypa/warehouse/issues/7119). As far as the comment goes, I'm not sure on what might be better... I just realized when I read it that it sent me to go look and see what the `tag` column on the other events held and I came up without a goose.

**Comment by xmunoz** [comment]
I think this test failure is flaky. Could someone re-run it for me?

File: warehouse/malware/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (xmunoz) [reply warehouse/malware/models.py]: FYI while ID is unique, it was not specified as a primary key. I've added at in the followup commit.

[the remark this answers is not in the document]
Comment (ewdurbin) [reply warehouse/malware/models.py]: Models that inherit from `db.Model` inherit a unique UUID `id` field [set as a primary key](https://github.com/pypa/warehouse/blob/de6d014f62021cb915187f75458fa3ce53d115a7/warehouse/db.py#L109), so I'm not sure what needs to change.

[the remark this answers is not in the document]
Comment (xmunoz) [reply warehouse/malware/models.py]: That's what I thought, when I ran the migration I got this error:
`sqlalchemy.exc.ProgrammingError: (psycopg2.errors.InvalidForeignKey) there is no unique constraint matching given keys for referenced table "release_files"`
I then manually checked the db:
`
warehouse=# \d release_files;
...
    "release_files_release_id_idx" btree (release_id)
`
This is what prompted me to add the primary key constraint.

[the remark this answers is not in the document]
Comment (ewdurbin) [reply warehouse/malware/models.py]: Huh, that seems wrong. Something must have been mishandled when we migrated to UUID ids in #4958. I'll file an issue.

File: warehouse/migrations/versions/061ff3d24c22_add_malware_detection_tables.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Add malware detection tables
+
+Revision ID: 061ff3d24c22
+Revises: 08aedc089eaf
```
Comment (ewdurbin) [inline warehouse/migrations/versions/061ff3d24c22_add_malware_detection_tables.py]: Update to `b5bb5d08543d`

File: warehouse/migrations/versions/061ff3d24c22_add_malware_detection_tables.py [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+Create Date: 2019-12-18 17:27:00.183542
+"""
+import citext
+import sqlalchemy as sa
+
+from alembic import op
+from sqlalchemy.dialects import postgresql
+
+revision = "061ff3d24c22"
+down_revision = "08aedc089eaf"
```
Comment (ewdurbin) [inline warehouse/migrations/versions/061ff3d24c22_add_malware_detection_tables.py]: Update to `b5bb5d08543d`

**Pull request merged**